### ISSUE TITLE
A ItemDetailsBLOC was used after being disposed. Once you have called dispose() on a ItemDetailsBLOC, it can no longer be used.

### ISSUE THREAD

**Issue by BasedMusa** [rrousselGit/provider#316]
Hi, In my e-commerce app, I have to load the details for each item individually, but I want the process to feel seamless, so I created my own ItemDetailsBLOC model, and initiated it in my grid view item. So that when, that it starts loading data as soon as the item is rendered, if the user opens any of the item's detail page, the data will already be there or will at least loading will feel shorter.

I get the following error, when I scroll really fast through the GridView.

```
ItemDetailsBLOC: LoadData: UnexpectedError: A ItemDetailsBLOC was used after being disposed.
Once you have called dispose() on a ItemDetailsBLOC, it can no longer be used.
[VERBOSE-2:ui_dart_state.cc(157)] Unhandled Exception: A ItemDetailsBLOC was used after being disposed.
Once you have called dispose() on a ItemDetailsBLOC, it can no longer be used.
#0      ChangeNotifier._debugAssertNotDisposed.<anonymous closure> (package:flutter/src/foundation/change_notifier.dart:105:9)
#1      ChangeNotifier._debugAssertNotDisposed (package:flutter/src/foundation/change_notifier.dart:111:6)
#2      ChangeNotifier.notifyListeners (package:flutter/src/foundation/change_notifier.dart:200:12)
#3      ItemDetailsBLOC.taskStatus= (package:mega_dot_pk/blocs/item_details_bloc.dart:21:5)
#4      ItemDetailsBLOC.loadData (package:mega_dot_pk/blocs/item_details_bloc.dart:40:7)
<asynchronous suspension>
#5      new ItemDetailsBLOC (package:mega_dot_pk/blocs/item_details_bloc.dart:26:5)
#6      _ItemGridItemState.build.<anonymous closure> (package:mega_dot_pk/widgets/item_grid_item.dart:27:24)
#7      BuilderStateDelegate.initDelegate (package:provider/src/delegate_widget.dart:249:14)
#8   <…>
Below is how I create and pass my ItemDetailsBLOC to the next page. Note, I am passing the BLOC because when I call Navigator.push, the Consumer will not be able to directly fetch this because it's not in the same widget tree.
```

GridItem (Stateful)
```Dart
@override
  Widget build(BuildContext context) => ChangeNotifierProvider(
          create: (_) => ItemDetailsBLOC(widget.item),
          child: Consumer<ItemDetailsBLOC>(
            builder: (_, bloc, __) => InkWell(
              onTap: () => _goToDetails(bloc),
              child: _myGridItem(),
           ),
         ),
       );
```

ItemDetailsPage (Stateful)
```Dart
class ItemDetailPage extends StatefulWidget {
  final ItemDetailsBLOC bloc;

  ItemDetailPage(this.bloc);
  @override
  _ItemDetailPageState createState() => _ItemDetailPageState();
}

class _ItemDetailPageState extends State<ItemDetailPage> {

 @override
  Widget build(BuildContext context) => Scaffold(
        appBar: _appBar(),
        body: _body(widget.bloc),
        bottomNavigationBar: _bottomAppBar(),
        backgroundColor: Theme.of(context).canvasColor,
      );
```

**Comment by rrousselGit** [comment]
I do not have enough information to reproduce this.

Consider making a small example showcasing the problem

**Comment by BasedMusa** [comment]
I'm sorry but I unfortunately won't be able to share a complete example since it's a HUGE codebase, and the app has a lot of self dependencies such as a complete Auth system. But I can try explaining my issue further.

I have a page called `ItemsGridPage`. Inside that page I am using `ItemsBLOC` to manage a list of products from an e-commerce site. The build method is like so: 
```Dart
@override
  Widget build(BuildContext context) => ChangeNotifierProvider(
        create: (_) => ItemsBLOC(widget.category, Provider.of<AuthenticationProviderBLOC>(context).user.uid),
        child: Scaffold(
          body: _body(),
        ),
      );

_body() => Consumer<ItemsBLOC>(
        builder: (_, bloc, __) => NestedScrollView(
          headerSliverBuilder: (BuildContext context, bool innerBoxScrolled) =>
              [_appBar(bloc.hasItems)],
          controller: _headerScrollController,
          body: Container(
            child: bloc.taskStatus.loading
                ? BrandedLoadingIndicator()
                : bloc.taskStatus.error
                    ? BrandedErrorPage(bloc.taskStatus, bloc.loadData)
                    : bloc.hasItems && bloc.taskStatus.loading == false
                        ? _ItemsGridView(bloc)
                        : Text(
                            "No Products Found",
                          ),
          ),
        ),
      );
```

`ItemsGridView` builds it's grid items/children like this:

```Dart
SliverChildBuilderDelegate(
    (context, i) => ItemGridItem(i, widget.bloc.items[i]),
    childCount: widget.bloc.items?.length ?? 
),
```

This is `ItemGridItem` `s build method:
```Dart
@override
  Widget build(BuildContext context) => ChangeNotifierProvider(
          create: (_) => ItemDetailsBLOC(widget.item),
          child: Consumer<ItemDetailsBLOC>(
            builder: (_, bloc, __) => InkWell(
              onTap: () => _goToDetails(bloc),
              child: _myGridItem(),
           ),
         ),
       );

void _goToDetails(ItemDetailsBLOC bloc) =>
    Navigator.push(
      context,
      MaterialPageRoute(
        builder: (context) => ItemDetailPage(widget.item, bloc),
      ),
);
```

This is my `ItemDetailsPage`, you go here when you tap on `ItemGridItem` in the GridView.
```Dart
class ItemDetailPage extends StatefulWidget {
  final ItemDetailsBLOC bloc;

  ItemDetailPage(this.bloc);
  @override
  _ItemDetailPageState createState() => _ItemDetailPageState();
}

class _ItemDetailPageState extends State<ItemDetailPage> {

 @override
  Widget build(BuildContext context) => Scaffold(
        appBar: _appBar(),
        body: _body(widget.bloc),
        bottomNavigationBar: _bottomAppBar(),
        backgroundColor: Theme.of(context).canvasColor,
      );
```

**WHAT I WANT TO DO**

Basically, I have 100s of Providers in a GridView and when I tap on a GridViewItem (`ItemGridItem` widget), it should take me to `ItemDetailsPage` and also pass the Provider the new page. This functionality works fine overall, I am using it even now. But the issue is that when I scroll really fast through a `GridView.builder`, it starts showing the above messages, I'm kind of confused, so I would really appreciate help on this, and I hope this helps you understand, thank you in advance :).

**Comment by rrousselGit** [comment]
> I'm sorry but I unfortunately won't be able to share a complete example since it's a HUGE codebase

Build a small example from scratch to reproduce the error
I cannot help you otherwise.

**Comment by rrousselGit** [comment]
Considering the lack of response and the age of the issue, I'll close this for now.

If you still face that problem, please add a minimalist way to reproduce the error.
